File: backend/app/schemas/mcp.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


CanonicalTransport = Literal["stdio", "streamable-http", "sse"]
McpManagementState = Literal[
    "disabled",
    "not_configured",
    "connecting",
    "ready",
    "degraded",
    "error",
    "stale",
]


class McpConnectionModel(BaseModel):
    type: CanonicalTransport = "stdio"
    url: str | None = None
    headers: dict[str, str] = Field(default_factory=dict)
    command: str | None = None
    args: list[str] = Field(default_factory=list)
    env: dict[str, str] = Field(default_factory=dict)


class McpAuthModel(BaseModel):
    required_env: list[str] = Field(default_factory=list)


class McpTimeoutsModel(BaseModel):
    connect: float = 60.0
    call: float = 25.0


class McpToolsModel(BaseModel):
    include: list[str] = Field(default_factory=list)
    exclude: list[str] = Field(default_factory=list)
    resources: bool = False
    prompts: bool = False


class McpRuntimeModel(BaseModel):
    supports_parallel_tool_calls: bool = False


class McpServerWrite(BaseModel):
    """创建/更新请求体（可含掩码字段）。"""

    display_name: str | None = None
    enabled: bool = True
    connection: McpConnectionModel = Field(default_factory=McpConnectionModel)
    auth: McpAuthModel = Field(default_factory=McpAuthModel)
    timeouts: McpTimeoutsModel = Field(default_factory=McpTimeoutsModel)
    tools: McpToolsModel = Field(default_factory=McpToolsModel)
    runtime: McpRuntimeModel = Field(default_factory=McpRuntimeModel)
    clear_headers: list[str] = Field(default_factory=list)
    clear_env: list[str] = Field(default_factory=list)


class McpServerDraft(McpServerWrite):
    """校验/测试草稿，可带临时 id。"""

    id: str | None = None


class McpStatusView(BaseModel):
    state: McpManagementState
    connected: bool = False
    tool_count: int = 0
    tool_names: list[str] = Field(default_factory=list)
    error: str | None = None
    error_code: str | None = None
    captured_at: float | None = None
    source: str | None = None
    stale: bool = False
    config_revision: int = 0
    applied_revision: int = 0


class McpServerListItem(BaseModel):
    id: str
    display_name: str
    enabled: bool
    transport: CanonicalTransport
    state: McpManagementState
    tool_count: int = 0
    last_checked_at: float | None = None
    error_summary: str | None = None
    can_connect: bool = False
    validation_errors: list[str] = Field(default_factory=list)
    unresolved_variables: list[str] = Field(default_factory=list)
    source: Literal["mcp", "mcp_servers"] = "mcp"


class McpServerDetail(McpServerListItem):
    connection: McpConnectionModel
    auth: McpAuthModel = Field(default_factory=McpAuthModel)
    timeouts: McpTimeoutsModel = Field(default_factory=McpTimeoutsModel)
    tools: McpToolsModel = Field(default_factory=McpToolsModel)
    runtime: McpRuntimeModel = Field(default_factory=McpRuntimeModel)
    warnings: list[str] = Field(default_factory=list)
    unknown_fields: dict[str, Any] = Field(default_factory=dict)
    status: McpStatusView | None = None
    header_refs: dict[str, str] = Field(default_factory=dict)
    env_refs: dict[str, str] = Field(default_factory=dict)


class McpListResponse(BaseModel):
    servers: list[McpServerListItem]
    total: int
    config_revision: int
    warnings: list[str] = Field(default_factory=list)
    legacy_shadowed: bool = False
    source: Literal["mcp", "mcp_servers", "empty"] = "empty"


class McpFieldError(BaseModel):
    path: str
    kind: str
    message: str
    expected: Any | None = None


class McpValidateResponse(BaseModel):
    ok: bool
    validation_errors: list[str] = Field(default_factory=list)
    errors: list[McpFieldError] = Field(default_factory=list)
    unresolved_variables: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    can_connect: bool = False
    normalized: dict[str, Any] = Field(default_factory=dict)


class McpTestResponse(BaseModel):
    ok: bool
    state: McpManagementState
    duration_ms: float
    tool_count: int = 0
    tool_names: list[str] = Field(default_factory=list)
    error: str | None = None
    error_code: str | None = None
    replaced_runtime: bool = False


class McpRefreshServerResult(BaseModel):
    id: str
    ok: bool
    state: McpManagementState
    tool_count: int = 0
    error: str | None = None
    error_code: str | None = None


class McpRefreshResponse(BaseModel):
    results: list[McpRefreshServerResult]
    config_revision: int
    applied_revision: int


class McpMutationResponse(BaseModel):
    status: str = "ok"
    server: McpServerDetail
    config_revision: int

File: frontend/src/api/mcp.ts
import api from './index'
import type { ConfigFieldError } from './config'
import {
  MCP_SECRET_MASK,
  type McpAuth,
  type McpConnection,
  type McpRuntime,
  type McpServerJsonDraft,
  type McpServerWrite,
  type McpTimeouts,
  type McpToolsFilter,
  type McpTransport,
} from '@/config/types'

export type McpState =
  | 'disabled'
  | 'not_configured'
  | 'connecting'
  | 'ready'
  | 'degraded'
  | 'error'
  | 'stale'

export type {
  McpTransport,
  McpConnection,
  McpAuth,
  McpTimeouts,
  McpToolsFilter,
  McpRuntime,
  McpServerJsonDraft,
  McpServerWrite,
}
export { MCP_SECRET_MASK }

export interface McpServerListItem {
  id: string
  display_name: string
  enabled: boolean
  transport: McpTransport
  state: McpState
  tool_count: number
  last_checked_at?: number | null
  error_summary?: string | null
  can_connect: boolean
  validation_errors: string[]
  unresolved_variables: string[]
  source: 'mcp' | 'mcp_servers'
}

export interface McpServerDetail extends McpServerListItem {
  connection: McpConnection
  auth: McpAuth
  timeouts: McpTimeouts
  tools: McpToolsFilter
  runtime: McpRuntime
  warnings: string[]
  unknown_fields: Record<string, unknown>
  status?: {
    state: McpState
    connected: boolean
    tool_count: number
    tool_names: string[]
    error?: string | null
    error_code?: string | null
    captured_at?: number | null
    source?: string | null
    stale: boolean
    config_revision: number
    applied_revision: number
  } | null
  header_refs: Record<string, string>
  env_refs: Record<string, string>
}

export interface McpListResponse {
  servers: McpServerListItem[]
  total: number
  config_revision: number
  warnings: string[]
  legacy_shadowed: boolean
  source: 'mcp' | 'mcp_servers' | 'empty'
}

export interface McpValidateResponse {
  ok: boolean
  validation_errors: string[]
  unresolved_variables: string[]
  warnings: string[]
  can_connect: boolean
  normalized: Record<string, unknown>
  errors?: ConfigFieldError[]
}

export interface McpValidationErrorBody {
  detail?: string | { validation_errors?: string[]; message?: string }
  code?: string
  message?: string
  validation_errors?: string[]
  errors?: ConfigFieldError[]
}

export interface McpTestResponse {
  ok: boolean
  state: McpState
  duration_ms: number
  tool_count: number
  tool_names: string[]
  error?: string | null
  error_code?: string | null
  replaced_runtime: boolean
}

export interface McpRefreshResponse {
  results: Array<{
    id: string
    ok: boolean
    state: McpState
    tool_count: number
    error?: string | null
    error_code?: string | null
  }>
  config_revision: number
  applied_revision: number
}

export interface McpMutationResponse {
  status: string
  server: McpServerDetail
  config_revision: number
}

export const mcpApi = {
  list: () => api.get<McpListResponse>('/mcp/list'),
  get: (id: string) => api.get<McpServerDetail>(`/mcp/${encodeURIComponent(id)}`),
  upsert: (id: string, body: McpServerWrite) =>
    api.put<McpMutationResponse>(`/mcp/${encodeURIComponent(id)}`, body),
  remove: (id: string) => api.delete<{ status: string; id: string; config_revision: number }>(
    `/mcp/${encodeURIComponent(id)}`,
  ),
  enable: (id: string) => api.post<McpMutationResponse>(`/mcp/${encodeURIComponent(id)}/enable`),
  disable: (id: string) => api.post<McpMutationResponse>(`/mcp/${encodeURIComponent(id)}/disable`),
  validate: (body: McpServerWrite & { id?: string }) =>
    api.post<McpValidateResponse>('/mcp/validate', body),
  test: (body: McpServerWrite & { id?: string }) =>
    api.post<McpTestResponse>('/mcp/test', body),
  refresh: () => api.post<McpRefreshResponse>('/mcp/refresh'),
}
